core/version_api.php

```diff
@@ -530,7 +530,7 @@ function version_get_all_rows( $p_project_ids, $p_released = null, $p_obsolete =
 		if( !empty( $g_cache_versions_project[$t_project_id]) ) {
 			foreach( $g_cache_versions_project[$t_project_id] as $t_id ) {
 				$t_version_row = version_cache_row( $t_id );
-				if( $p_obsolete == false && (int)$t_version_row['obsolete'] == 1 ) {
+				if( $p_obsolete === false && (int)$t_version_row['obsolete'] == 1 ) {
```

Comment:
Didn't test, but isn't the real problem that there are places where `version_get_all_rows` is called with `null` for parameter `$p_obsolete`, but should be `true` or `false`?
```
 * @param boolean $p_obsolete   Whether to include obsolete versions.
```
e.g. in `manage_proj_edit_page.php` there is
```
$t_versions = version_get_all_rows( $f_project_id, null, null );
```

Reply:
$p_released is a 3 way selector (i have updated the doc):
```php
 @param boolean $p_released   Whether to show only released, unreleased, or both.
                  For this parameter, use contants defined as:
                  VERSION_ALL (null): returns any
                  VERSION_FUTURE (false): returns only unreleased versions
                  VERSION_RELEASED (true): returns only released versions
```
In the previous implementation, $p_obsolete worked in the same way:
```php
	if( $p_obsolete !== null ) {
		$t_query .= ' AND obsolete = ' . db_param();
		$t_query_params[] = (bool)$p_obsolete;
	}
```
But having a default value of `false`, it was never used with a value of `true` (meaning, return only obsolete versions)
After rewriting this function, now it's a true/false parameter meaning precisely "include, or not, obsolete versions"


> but isn't the real problem that there are places where version_get_all_rows is called with null for parameter $p_obsolete, but should be true or false?

probably, those should be cleaned up to use true/false.
since this is a regression fix, i'd rather not include it in this PR

Reply:
@cproensa So what you're saying is that it is no longer possible to retrieve only obsolete versions ? OK, fine. 

Adding the strict-type check against *false* as you are proposing is a good thing to do as a safety net, but I would agree with @atrol that since the semantics of the `$p_obsolete` parameter have changed, then the **proper fix** for this regression, is to change the callers to use *true* instead of null. 
```
$ git grep version_get_all_rows\(.*,.*,\ null
api/soap/mc_project_api.php:839:    $t_versions = version_get_all_rows( $p_project_id, null, null );
core/print_api.php:865: $t_versions = version_get_all_rows( $t_project_ids, $p_released, null );
manage_proj_edit_page.php:495:  $t_versions = version_get_all_rows( $f_project_id, null, null );
```

What could be done outside of this PR if you prefer, is to make sure the callers are referencing the *VERSION_xxx* constants instead of true/false/null. That's only 5 occurences AFAICT
```
$ git grep -P '(?<!function )version_get_all_rows\(.*,\ (?!VERSION)'
api/soap/mc_project_api.php:839:    $t_versions = version_get_all_rows( $p_project_id, null, null );
changelog_page.php:248: $t_version_rows = version_get_all_rows( $t_project_id, null, false );
core/print_api.php:865: $t_versions = version_get_all_rows( $t_project_ids, $p_released, null );
core/version_api.php:576:   $t_versions = version_get_all_rows( $c_project_id, null /* released: any */, true /* incl. obsolete */ ,  $p_inherit );
manage_proj_edit_page.php:495:  $t_versions = version_get_all_rows( $f_project_id, null, null );
```

Reply:
@cproensa replied via chat
> 19-Apr-2019 11:10
> ok. for #1492 i agree with your suggestions, and will try to include them.

Since he's busy, I'll include them myself :wink: